spring-cloud-kubernetes-fabric8-discovery/src/test/java/org/springframework/cloud/kubernetes/fabric8/discovery/Fabric8EndpointsAndEndpointSlicesTests.java

```diff
@@ -0,0 +1,338 @@
+	 *
+	 *     We search with labels = {color = blue}
+	 *     Since namespaces = [namespaceA], we wil take podB, because all-namespace=false (podD is not part of the response)
+	 *
+	 * </pre>
+	 */
+	abstract void testAllNamespacesFalseNamespacesPresent();
+
+	/**
+	 * <pre>
+	 *     - all-namespaces = false
+	 *     - namespaces = []
+	 *
+	 *     - we have 5 pods involved in this test
+	 * 	   - podA in namespaceA with no labels
+	 * 	   - podB in namespaceA with labels {color=blue}
+	 * 	   - podC in namespaceA with labels {color=red}
+	 * 	   - podD in namespaceB with labels {color=blue}
+	 * 	   - podE in namespaceB with no labels
+	 *
+	 *     We search with labels = {color = blue}
+	 *     Since namespaces = [], we wil take podB, because all-namespace=false (podD is not part of the response)
+	 *
+	 * </pre>
+	 */
+	abstract void testAllNamespacesFalseNamespacesNotPresent();
+
+	/**
+	 * <pre>
+	 *     - all-namespaces = false
+	 *     - namespaces = [namespaceA, namespaceB]
+	 *
+	 *     - we have 7 pods involved in this test
+	 * 	   - podA in namespaceA with no labels
+	 * 	   - podB in namespaceA with labels {color=blue}
+	 * 	   - podC in namespaceA with labels {color=red}
+	 * 	   - podD in namespaceB with labels {color=blue}
+	 * 	   - podE in namespaceB with no labels
+	 * 	   - podF in namespaceB with labels {color=blue}
+	 * 	   - podO in namespaceC with labels {color=blue}
+	 *
+	 *     We search with labels = {color = blue}
+	 *     Since namespaces = [namespaceA, namespaceB], we wil take podB, podD and podF,
+	 *     but will not take podO
+	 *
+	 * </pre>
+	 */
+	abstract void testTwoNamespacesOutOfThree();
+
+	KubernetesCatalogWatch createWatcherInAllNamespacesWithLabels(Map<String, String> labels, Set<String> namespaces,
+			boolean endpointSlices) {
+
+		boolean allNamespaces = true;
+		KubernetesDiscoveryProperties properties = new KubernetesDiscoveryProperties(true, allNamespaces, namespaces,
+				true, 60, false, "", Set.of(), labels, "", null, 0, endpointSlices);
+		KubernetesCatalogWatch watch = new KubernetesCatalogWatch(mockClient(), properties, NAMESPACE_PROVIDER);
+
+		if (endpointSlices) {
+			watch = Mockito.spy(watch);
+			Mockito.doReturn(new Fabric8EndpointSliceV1CatalogWatch()).when(watch).stateGenerator();
+		}
+
+		watch.postConstruct();
+		watch.setApplicationEventPublisher(APPLICATION_EVENT_PUBLISHER);
+		return watch;
+
+	}
+
+	KubernetesCatalogWatch createWatcherInSpecificNamespaceWithLabels(String namespace, Map<String, String> labels,
+			boolean endpointSlices) {
+
+		when(NAMESPACE_PROVIDER.getNamespace()).thenReturn(namespace);
+
+		boolean allNamespaces = false;
+		KubernetesDiscoveryProperties properties = new KubernetesDiscoveryProperties(true, allNamespaces,
+				Set.of(namespace), true, 60, false, "", Set.of(), labels, "", null, 0, endpointSlices);
+		KubernetesCatalogWatch watch = new KubernetesCatalogWatch(mockClient(), properties, NAMESPACE_PROVIDER);
+
+		if (endpointSlices) {
```

Comment:
this is where I create a spy and mock just that method to return to me the EndpointSlices implementation.